modules/dag-shared/src/main/scala/org/tessellation/dag/block/BlockValidator.scala

```diff
@@ -0,0 +1,153 @@
+package org.tessellation.dag.block
+
+import cats.Applicative
+import cats.data.Validated.{Invalid, Valid}
+import cats.data.{NonEmptyList, ValidatedNel}
+import cats.effect.Async
+import cats.syntax.applicative._
+import cats.syntax.flatMap._
+import cats.syntax.foldable._
+import cats.syntax.functor._
+import cats.syntax.show._
+import cats.syntax.traverse._
+import cats.syntax.validated._
+
+import scala.util.control.NoStackTrace
+
+import org.tessellation.dag.block.BlockValidator._
+import org.tessellation.dag.transaction.TransactionValidator
+import org.tessellation.dag.transaction.TransactionValidator.TransactionValidationError
+import org.tessellation.dag.transaction.filter.Consecutive
+import org.tessellation.dag.types.{BlockReference, DAGBlock}
+import org.tessellation.kryo.KryoSerializer
+import org.tessellation.schema.address.Address
+import org.tessellation.schema.transaction.{Transaction, TransactionReference}
+import org.tessellation.security.SecurityProvider
+import org.tessellation.security.signature.Signed
+
+abstract class BlockValidator[F[_]](transactionValidator: TransactionValidator[F], consensusPeersCount: Int) {
+
+  implicit val F: Async[F]
+  implicit val kryoSerializer: KryoSerializer[F]
+  implicit val securityProvider: SecurityProvider[F]
+
+  def areParentsAccepted(block: DAGBlock): F[Map[BlockReference, Boolean]]
+
+  def getLastAcceptedTransactionRef(address: Address): F[TransactionReference]
+
+  def validate(signedBlock: Signed[DAGBlock]): F[ValidationResult[Signed[DAGBlock]]] =
+    for {
+      signaturesValidation <- validateSignatures(signedBlock)
+      signaturesCountValidation = validateSignaturesCount(signedBlock)
+      parentsValidation <- validateParents(signedBlock)
+      transactionsValidation <- validateTransactions(signedBlock)
+    } yield
+      signaturesValidation
+        .product(signaturesCountValidation)
+        .product(parentsValidation)
+        .product(transactionsValidation)
+        .map(_ => signedBlock)
+
+  private def validateSignatures(signedBlock: Signed[DAGBlock]): F[ValidationResult[Signed[DAGBlock]]] =
+    signedBlock.hasValidSignature
+      .ifM(
+        signedBlock.validNel.pure[F],
+        Applicative[F].pure(InvalidBlockSignatures.invalidNel)
+      )
+
+  private def validateSignaturesCount(signedBlock: Signed[DAGBlock]): ValidationResult[Signed[DAGBlock]] = {
+    val uniqueSignersCount = signedBlock.proofs.map(_.id).toList.toSet.size
+
+    // TODO: is more signatures than expected ok?
```

Comment:
more than expected is fine, although such a case should not happen for now.

Reply:
Yeah I wonder if we should just go with equal then? Especially that it's configurable here.

Reply:
I don't think so. IMO greater or equal is 100% correct